File: src/sys/dotnet/fan/sys/Range.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   29 Jan 07  Andy Frank  Creation
//

using System;
using System.Globalization;

namespace Fan.Sys
{
  /// <summary>
  /// Range.
  /// </summary>
  public sealed class Range : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructors
  //////////////////////////////////////////////////////////////////////////

    public static Range makeInclusive(long start, long end)
    {
      return new Range(start, end, false);
    }

    public static Range makeExclusive(long start, long end)
    {
      return new Range(start, end, true);
    }

    public static Range make(long start, long end, bool exclusive)
    {
      return new Range(start, end, exclusive);
    }

    public static Range fromStr(string s) { return fromStr(s, true); }
    public static Range fromStr(string s, bool check)
    {
      try
      {
        int dot = s.IndexOf('.');
        if (s[dot+1] != '.') throw new Exception();
        bool exclusive = s[dot+2] == '<';
        long start = Convert.ToInt64(s.Substring(0, dot));
        long end   = Convert.ToInt64(s.Substring(dot + (exclusive?3:2)));
        return new Range(start, end, exclusive);
      }
      catch (Exception) {}
      if (!check) return null;
      throw ParseErr.make("Range", s).val;
    }

    private Range(long start, long end, bool exclusive)
    {
      this.m_start = start;
      this.m_end = end;
      this.m_exclusive = exclusive;
    }

  //////////////////////////////////////////////////////////////////////////
  // Range
  //////////////////////////////////////////////////////////////////////////

    public long start()
    {
      return m_start;
    }

    public long end()
    {
      return m_end;
    }

    public bool inclusive()
    {
      return !m_exclusive;
    }

    public bool exclusive()
    {
      return m_exclusive;
    }

    public bool isEmpty()
    {
      return m_exclusive && m_start == m_end;
    }

    public Long min()
    {
      if (isEmpty()) return null;
      if (m_end < m_start) return Long.valueOf(m_exclusive ? m_end+1 : m_end);
      return Long.valueOf(m_start);
    }

    public Long max()
    {
      if (isEmpty()) return null;
      if (m_end < m_start) return Long.valueOf(m_start);
      return Long.valueOf(m_exclusive ? m_end-1 : m_end);
    }

    public Long first()
    {
      if (isEmpty()) return null;
      return Long.valueOf(m_start);
    }

    public Long last()
    {
      if (isEmpty()) return null;
      if (!m_exclusive) return Long.valueOf(m_end);
      if (m_start < m_end) return Long.valueOf(m_end-1);
      return Long.valueOf(m_end+1);
    }

    public Range offset(long offset)
    {
      if (offset == 0) return this;
      return new Range(m_start+offset, m_end+offset, m_exclusive);
    }

    public bool contains(long i)
    {
      if (m_start < m_end)
      {
        if (m_exclusive)
          return m_start <= i && i < m_end;
        else
          return m_start <= i && i <= m_end;
      }
      else
      {
        if (m_exclusive)
          return m_end < i && i <= m_start;
        else
          return m_end <= i && i <= m_start;
      }
    }

    public void each(Func f)
    {
      long start = m_start;
      long end = m_end;
      if (start < end)
      {
        if (m_exclusive) --end;
        for (long i=start; i<=end; ++i) f.call(i);
      }
      else
      {
        if (m_exclusive) ++end;
        for (long i=start; i>=end; --i) f.call(i);
      }
    }

    public List map(Func f)
    {
      long start = m_start;
      long end = m_end;
      Type r = f.returns();
      if (r == Sys.VoidType) r = Sys.ObjType.toNullable();
      List acc = new List(r);
      if (start < end)
      {
        if (m_exclusive) --end;
        for (long i=start; i<=end; ++i) acc.add(f.call(i));
      }
      else
      {
        if (m_exclusive) ++end;
        for (long i=start; i>=end; --i) acc.add(f.call(i));
      }
      return acc;
    }

    public List toList()
    {
      int start = (int)m_start;
      int end = (int)m_end;
      List acc = new List(Sys.IntType);
      if (start < end)
      {
        if (m_exclusive) --end;
        acc.capacity(end-start+1);
        for (int i=start; i<=end; ++i)
          acc.add(Long.valueOf(i));
      }
      else
      {
        if (m_exclusive) ++end;
        acc.capacity(start-end+1);
        for (int i=start; i>=end; --i)
          acc.add(Long.valueOf(i));
      }
      return acc;
    }

    public long random()
    {
      return FanInt.random(this);
    }

    public override bool Equals(object obj)
    {
      if (obj is Range)
      {
        Range that = (Range)obj;
        return this.m_start == that.m_start &&
               this.m_end == that.m_end &&
               this.m_exclusive == that.m_exclusive;
      }
      return false;
    }

    public override int GetHashCode()
    {
      return m_start.GetHashCode() ^ m_end.GetHashCode();
    }

    public override long hash()
    {
      return m_start ^ m_end;
    }

    public override string toStr()
    {
      if (m_exclusive)
        return m_start.ToString() + "..<" + m_end.ToString();
      else
        return m_start.ToString() + ".." + m_end.ToString();
    }

    public override Type @typeof() { return Sys.RangeType; }

  //////////////////////////////////////////////////////////////////////////
  // Slice Utils
  //////////////////////////////////////////////////////////////////////////

    internal int start(int size)
    {
      int x = (int)m_start;
      if (x < 0) x = size + x;
      if (x > size) throw IndexErr.make(this).val;
      return x;
    }

    internal long start(long size)
    {
      long x = m_start;
      if (x < 0) x = size + x;
      if (x > size) throw IndexErr.make(this).val;
      return x;
    }

    internal int end(int size)
    {
      int x = (int)m_end;
      if (x < 0) x = size + x;
      if (m_exclusive) x--;
      if (x >= size) throw IndexErr.make(this).val;
      return x;
    }

    internal long end(long size)
    {
      long x = m_end;
      if (x < 0) x = size + x;
      if (m_exclusive) x--;
      if (x >= size) throw IndexErr.make(this).val;
      return x;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private long m_start, m_end;
    private bool m_exclusive;
  }
}

File: src/sys/java/fan/sys/Range.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   11 Dec 05  Brian Frank  Creation
//
package fan.sys;

/**
 * Range
 */
public final class Range
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Constructors
//////////////////////////////////////////////////////////////////////////

  public static Range makeInclusive(long start, long end)
  {
    return new Range(start, end, false);
  }

  public static Range makeExclusive(long start, long end)
  {
    return new Range(start, end, true);
  }

  public static Range make(long start, long end, boolean exclusive)
  {
    return new Range(start, end, exclusive);
  }

  public static Range fromStr(String s) { return fromStr(s, true); }
  public static Range fromStr(String s, boolean checked)
  {
    try
    {
      int dot = s.indexOf('.');
      if (s.charAt(dot+1) != '.') throw new Exception();
      boolean exclusive = s.charAt(dot+2) == '<';
      long start = Long.parseLong(s.substring(0, dot));
      long end   = Long.parseLong(s.substring(dot + (exclusive?3:2)));
      return new Range(start, end, exclusive);
    }
    catch (Exception e) {}
    if (!checked) return null;
    throw ParseErr.make("Range", s);
  }

  private Range(long start, long end, boolean exclusive)
  {
    this.start = start;
    this.end = end;
    this.exclusive = exclusive;
  }

//////////////////////////////////////////////////////////////////////////
// Range
//////////////////////////////////////////////////////////////////////////

  public final long start()
  {
    return start;
  }

  public final long end()
  {
    return end;
  }

  public final boolean inclusive()
  {
    return !exclusive;
  }

  public final boolean exclusive()
  {
    return exclusive;
  }

  public final boolean isEmpty()
  {
    return exclusive && start == end;
  }

  public final Long min()
  {
    if (isEmpty()) return null;
    if (end < start) return Long.valueOf(exclusive ? end+1 : end);
    return Long.valueOf(start);
  }

  public final Long max()
  {
    if (isEmpty()) return null;
    if (end < start) return Long.valueOf(start);
    return Long.valueOf(exclusive ? end-1 : end);
  }

  public final Long first()
  {
    if (isEmpty()) return null;
    return Long.valueOf(start);
  }

  public final Long last()
  {
    if (isEmpty()) return null;
    if (!exclusive) return Long.valueOf(end);
    if (start < end) return Long.valueOf(end-1);
    return Long.valueOf(end+1);
  }

  public final boolean contains(long i)
  {
    if (start < end)
    {
      if (exclusive)
        return start <= i && i < end;
      else
        return start <= i && i <= end;
    }
    else
    {
      if (exclusive)
        return end < i && i <= start;
      else
        return end <= i && i <= start;
    }
  }

  public final Range offset(long offset)
  {
    if (offset == 0) return this;
    return new Range(start+offset, end+offset, exclusive);
  }

  public final void each(Func f)
  {
    long start = this.start;
    long end = this.end;
    if (start < end)
    {
      if (exclusive) --end;
      for (long i=start; i<=end; ++i) f.call(Long.valueOf(i));
    }
    else
    {
      if (exclusive) ++end;
      for (long i=start; i>=end; --i) f.call(Long.valueOf(i));
    }
  }

  public final List map(Func f)
  {
    long start = this.start;
    long end = this.end;
    Type r = f.returns();
    if (r == Sys.VoidType) r = Sys.ObjType.toNullable();
    List acc = new List(r);
    if (start < end)
    {
      if (exclusive) --end;
      for (long i=start; i<=end; ++i) acc.add(f.call(Long.valueOf(i)));
    }
    else
    {
      if (exclusive) ++end;
      for (long i=start; i>=end; --i) acc.add(f.call(Long.valueOf(i)));
    }
    return acc;
  }

  public final List toList()
  {
    long start = this.start;
    long end = this.end;
    List acc = new List(Sys.IntType);
    if (start < end)
    {
      if (exclusive) --end;
      acc.capacity(Long.valueOf(end-start+1));
      for (long i=start; i<=end; ++i) acc.add(Long.valueOf(i));
    }
    else
    {
      if (exclusive) ++end;
      acc.capacity(Long.valueOf(start-end+1));
      for (long i=start; i>=end; --i) acc.add(Long.valueOf(i));
    }
    return acc;
  }

  public final long random()
  {
    return FanInt.random(this);
  }

  public final boolean equals(Object object)
  {
    if (object instanceof Range)
    {
      Range that = (Range)object;
      return this.start == that.start &&
             this.end == that.end &&
             this.exclusive == that.exclusive;
    }
    return false;
  }

  public final long hash()
  {
    return (start << 24) ^ end;
  }

  public String toStr()
  {
    if (exclusive)
      return start + "..<" + end;
    else
      return start + ".." + end;
  }

  public Type typeof() { return Sys.RangeType; }

//////////////////////////////////////////////////////////////////////////
// Slice Utils
//////////////////////////////////////////////////////////////////////////

  final int start(int size)
  {
    int x = (int)start;
    if (x < 0) x = size + x;
    if (x > size) throw IndexErr.make(this);
    return x;
  }

  final long start(long size)
  {
    long x = start;
    if (x < 0) x = size + x;
    if (x > size) throw IndexErr.make(this);
    return x;
  }

  final int end(int size)
  {
    int x = (int)end;
    if (x < 0) x = size + x;
    if (exclusive) x--;
    if (x >= size) throw IndexErr.make(this);
    return x;
  }

  final long end(long size)
  {
    long x = end;
    if (x < 0) x = size + x;
    if (exclusive) x--;
    if (x >= size) throw IndexErr.make(this);
    return x;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private final long start, end;
  private final boolean exclusive;
}